device/core/src/sak_authentication_provider.ts

```diff
@@ -0,0 +1,130 @@
+    if (this._shouldRenewToken()) {
+      this._renewToken();
+    }
+
+    /*Codes_SRS_NODE_SAK_AUTH_PROVIDER_16_003: [The `getDeviceCredentials` should call its callback with a `null` first parameter and a `DeviceCredentials` object as a second parameter, containing the latest valid token it generated.]*/
+    callback(null, this._credentials);
+  }
+
+  /**
+   * Sets the current security token for the device. Note that it will be overwritten by the timer that is running in the background.
+   * If you're looking for an AuthenticationProvider where you maintain the lifecycle of the token yourself, you should look at the `SharedAccessSignatureAuthorizationProvider` class.
+   *
+   * @param sharedAccessSignature shared access signature that should be stored.
+   */
+  updateSharedAccessSignature(sharedAccessSignature: string): void {
+    /*Codes_SRS_NODE_SAK_AUTH_PROVIDER_16_004: [The `updateSharedAccessSignature` method shall save the `sharedAccessSignature` passed as a parameter and return it when `getDeviceCredentials` is called, until it gets renewed.]*/
+    debug('Manually replacing a shared access signature when automatic renewal is active');
+    this._credentials.sharedAccessSignature = sharedAccessSignature;
+  }
+
+  private _shouldRenewToken(): boolean {
+    const currentTimeInSeconds = Math.floor(Date.now() / 1000);
+    return (this._currentTokenExpiryTimeInSeconds - currentTimeInSeconds) < this._tokenRenewalMarginInSeconds;
+  }
+
+  private _renewToken(): void {
+    if (this._renewalTimeout) {
+      clearTimeout(this._renewalTimeout);
+    }
+
+    /*Codes_SRS_NODE_SAK_AUTH_PROVIDER_16_009: [Every token shall be created with a validity period of `tokenValidTimeInSeconds` if specified when the constructor was called, or 1 hour by default.]*/
+    const newExpiry =  Math.floor(Date.now() / 1000) + this._tokenValidTimeInSeconds;
+
+    /*Codes_SRS_NODE_SAK_AUTH_PROVIDER_16_010: [Every token shall be created using the `azure-iot-common.SharedAccessSignature.create` method and then serialized as a string, with the arguments to the create methods being:
+    ```
+    resourceUri: <IoT hub host>/devices/<deviceId>
+    keyName: the `SharedAccessKeyName` parameter of the connection string or `null`
+    key: the `SharedAccessKey` parameter of the connection string
+    expiry: the expiration time of the token, which is now + the token validity time, formatted as the number of seconds since Epoch (Jan 1st, 1970, 00:00 UTC).
+    ```]*/
+    const resourceUri = encodeUriComponentStrict(this._credentials.host + '/devices/' + this._credentials.deviceId);
+    const sas = SharedAccessSignature.create(resourceUri, this._credentials.sharedAccessKeyName, this._credentials.sharedAccessKey, newExpiry);
+    this._credentials.sharedAccessSignature = sas.toString();
+
+    const nextRenewalTimeout = (this._tokenValidTimeInSeconds - this._tokenRenewalMarginInSeconds) * 1000;
```

Comment:
>this._tokenValidTimeInSeconds - this._tokenRenewalMarginInSeconds [](start = 32, length = 65)

Suggest adding a throw  to the constructor if the margin time is greater than the valid time.

Reply:
good idea!

---
In reply to: [157022338](https://github.com/Azure/azure-iot-sdk-node/pull/179#discussion_r157022338) [](ancestors = 157022338)